### ISSUE TITLE
prefer-object-spread should ignore use with functions

### ISSUE THREAD

**Issue by davidgoli** [palantir/tslint#3055]
### Bug Report

- __TSLint version__: 5.4.3
- __TypeScript version__: 2.4.1
- __Running TSLint via__: (pick one) CLI

#### TypeScript code being linted

```ts
Object.assign(
  () => {},
  {
    foo() {},
  }
)
```

with `tslint.json` configuration:

```json
{
  "extends": ["tslint:latest", "tslint-config-prettier"],
  "rules": {
    "array-type": [true, "array"],
    "arrow-parens": [true, "ban-single-arg-parens"],
    "arrow-return-shorthand": [true, "multiline"],
    "await-promise": true,
    "curly": false,
    "interface-name": [true, "never-prefix"],
    "match-default-export-name": true,
    "max-classes-per-file": [true, 3],
    "max-line-length": [true, 120],
    "member-access": false,
    "no-consecutive-blank-lines": false,
    "no-duplicate-super": true,
    "no-invalid-template-strings": true,
    "no-misused-new": true,
    "no-sparse-arrays": true,
    "no-trailing-whitespace": false,
    "no-unnecessary-callback-wrapper": true,
    "no-unsafe-finally": true,
    "no-unused-expression": [true, "allow-fast-null-checks"],
    "object-literal-key-quotes": [true, "as-needed"],
    "object-literal-sort-keys": false,
    "ordered-imports": true,
    "prefer-const": true,
    "quotemark": false,
    "semicolon": false,
    "trailing-comma": false,
    "variable-name": [true, "ban-keywords", "check-format", "allow-leading-underscore", "allow-pascal-case"]
  }
}
```

#### Actual behavior

Result:
`ERROR: /Users/David/dev/hello-epics/power-up/src/app/cache.ts[12, 10]: 'Object.assign' returns the first argument. Prefer object spread if you want a new object.`

With prettier fix:
```ts
  {...() => {},
      foo() {},
  }
```

#### Expected behavior

prefer-object-spread should prefer `Object.assign` when the first argument is a function, as object rest spread on a function is somewhat nonsensical & probably not desired behavior.

**Comment by ajafff** [comment]
I don't know if we want to add special handling for functions (and classes and everything that has a call or contruct signature ...)
For the time being you can use the following workaround:

```ts
let fn = () => {};
Object.assign(fn, {
    foo() {}
});
```

**Comment by adidahiya** [comment]
The special handling could all be syntactic (done without the type checker), right? You'd just need to check for function expressions and class expressions as the first argument. I think we could take a PR for this.

**Comment by davidgoli** [comment]
I'll take a look at getting a PR together, thanks!

**Comment by davidgoli** [comment]
Closing, thanks for accepting my PR!
